python/ffsim/tenpy/circuits/gates.py

```diff
@@ -0,0 +1,460 @@
+    OSgate_sym = sym_cons_basis(OSgate)
+
+    return OSgate_sym
+
+
+def num_num_interaction(theta: float, spin: Spin) -> np.ndarray:
+    r"""The number-number interaction gate.
+
+    The number-number interaction gate as defined in
+    `apply_num_num_interaction <https://qiskit-community.github.io/ffsim/api/ffsim.html#ffsim.apply_num_num_interaction>`__,
+    returned in the TeNPy (N, Sz)-symmetry-conserved basis.
+
+    Args:
+        theta: The rotation angle.
+        spin: Choice of spin sector(s) to act on.
+
+            - To act on only spin alpha, pass :const:`ffsim.Spin.ALPHA`.
+            - To act on only spin beta, pass :const:`ffsim.Spin.BETA`.
+            - To act on both spin alpha and spin beta, pass
+              :const:`ffsim.Spin.ALPHA_AND_BETA` (this is the default value).
+
+    Returns:
+        The number-number interaction gate in the TeNPy (N, Sz)-symmetry-conserved
+        basis.
+    """
+
+    # alpha sector / up spins
+    if spin in [Spin.ALPHA, Spin.ALPHA_AND_BETA]:
+        # # Using TeNPy SpinHalfFermionSite(cons_N=None, cons_Sz=None) operators
+        # NNgate_a = sp.linalg.expm(1j * theta * np.kron(Nu, Nu))
+        NNgate_a = np.eye(16, dtype=complex)
+        for i in [5, 7, 13, 15]:
+            NNgate_a[i, i] = np.exp(1j * theta)
+
+    # beta sector / down spins
+    if spin in [Spin.BETA, Spin.ALPHA_AND_BETA]:
+        # # Using TeNPy SpinHalfFermionSite(cons_N=None, cons_Sz=None) operators
+        # NNgate_b = sp.linalg.expm(1j * theta * np.kron(Nd, Nd))
+        NNgate_b = np.eye(16, dtype=complex)
+        for i in [10, 11, 14, 15]:
+            NNgate_b[i, i] = np.exp(1j * theta)
+
+    # define total gate
+    if spin is Spin.ALPHA:
+        NNgate = NNgate_a
+    elif spin is Spin.BETA:
+        NNgate = NNgate_b
+    elif spin is Spin.ALPHA_AND_BETA:
+        NNgate = NNgate_a @ NNgate_b
+    else:
+        raise ValueError("undefined spin")
+
+    # convert to (N, Sz)-symmetry-conserved basis
+    NNgate_sym = sym_cons_basis(NNgate)
+
+    return NNgate_sym
+
+
+def apply_gate1(psi: MPS, U1: np.ndarray, site: int) -> None:
+    r"""Apply a single-site gate to a
+    `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+    wavefunction.
+
+    Args:
+        psi: The `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+            wavefunction.
+        U1: The single-site quantum gate.
+        site: The gate will be applied to `site` on the
+            `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+            wavefunction.
+
+    Returns:
+        None
+    """
+
+    # on-site
+    U1_npc = npc.Array.from_ndarray(U1, [shfs.leg, shfs.leg.conj()], labels=["p", "p*"])
+    psi.apply_local_op(site, U1_npc)
+
+
+def apply_gate2(
+    psi: MPS,
+    U2: np.ndarray,
+    site: int,
```

Comment:
Can we make this function accept both site indices, and raise an error if they are not adjacent? Then, if the site indices are not sorted, we can account for that by applying an appropriate permutation to U2, or something similar? If this is possible, then we can remote the unintuitive `conj` argument from `givens_rotation` above.

Reply:
Yes, this is possible. I think that it is as simple as transposing the gate.